### ISSUE TITLE
kfctl: kustomize component discovery logic is flawed

### ISSUE THREAD

**Issue by yanniszark** [kubeflow/kubeflow#3491]
`kfctl` uses a function called `mapDirs` to discover the components available and map each component's name to its path.
However, the logic it uses for discovery is flawed and can lead to perplexing errors.

https://github.com/kubeflow/kubeflow/blob/24c39e3cd66862b9433f23f7f9cc2937c70a39dd/bootstrap/v2/pkg/kfapp/kustomize/kustomize.go#L522-L523

In the above check, to check if a folder is a component, we check if the folder is a package name and a component name.
This makes the discovery's results _depend_ on what the user has asked.
The component discovery should yield the same results no matter what components the user asks for.

eg.
For a structure like:
```
|-- admission-webhook
    |-- bootstrap
        |-- base
    |-- webhook
        |-- base
```

If the KfDef is wrong like the following, the discovery code will think that the `admission-webhook` folder is a component, because of the check above.
```yaml
packages:
  - admission-webhook
components:
  - admission-webhook
```

If instead the KfDef is like the following, the components `bootstrap` and `webhook` will be discovered correctly.
```yaml
packages:
  - admission-webhook
components:
  - bootstrap
  - webhook
```

**Comment by jlewi** [comment]
Should we refactor the spec and get rid of components, packages, and componentParams?

This terminology no longer really makes sense now that we are using ksonnet.

Should we just refactor this into a list of applications

```
Applications:
   -  name: central-dashboard
      repoName:
      repoPath: 
      overlays:
         - overlay1 
         - overlay2
     parameters:
         - name: 
           value
``` 

For each application we can no specify its relative path within the repo so there's no more magic discovery going on.

**Comment by jlewi** [comment]
@yanniszark Can you help me understand what is release blocking? Unfortunately I think @kkasravi is OOO this rest of the week.

Why do we have packages and components? These concepts don't really make sense to me in the context of kustomize?

Should we pick one maybe packages and treat the value as a path to a kusotmization.yaml file.

So instead of 

```
packages:
  - admission-webhook
components:
  - bootstrap
  - webhook
```

We just do

```
components:
  - admission-webhook/bootstrap
  - admission-webhook/webhook
```

And treat the value as the relative path to the kustomization file?

Alternatively we could add something like application_types 

```
Applications:
   -  name: central-dashboard
      repoName:
      repoPath: 
      overlays:
         - overlay1 
         - overlay2
      parameters:
          - name: 
             value
```

And then add a method to try to auto populate Applications from packages/componentParams/componets perhaps by using the existing logic.

The advantage of this approach is that its much closer to a long term fix which decreases the likelihood that we run into other problems.

Providing a method to auto-populate Applications from packages/componentParams/componets means we don't have to update all existing usages in one go.

We can also write a unittest to verify the mapping happens correctly. 

Another advantage of this approach is that if need be we can just hardcode specific paths to use for existing components because any future changes should be using the new spec.

I think I favor this approach.

**Comment by yanniszark** [comment]
> Why do we have packages and components? These concepts don't really make sense to me in the context of kustomize?

Agreed

> Should we pick one maybe packages and treat the value as a path to a kusotmization.yaml file.

Do you mean the path to the directory of `kustomization.yaml`?
If yes, I agree.
Making the package references explicit will remove the (flawed) magic discovery and also provide namespaces for packages.
eg now you can't have two packages with the same name. With this approach you can, because they are separated by the parent path.

Regarding the application_types approach:
I believe this and fixing the package referencing are interconnected.
Adding a function to generate Applications from components + componentParams seems logical if we want to go towards Applications in the future.

My suggestion is:
1. Drop packages and fix components (or vice versa) so that they reference with a full path (with the root of the path being the manifest repo).
2. Add the auto-generate function for Applications.

In any case, I don't think we have anything to gain by keeping the current discovery logic.
It is easy to get wrong and also doesn't allow for namespacing of components.

**Comment by kkasravi** [comment]
great summary @yanniszark - thank you.

@jlewi - the original driver for this (flawed) logic was to keep compatibility with ksonnet. 
Since that isn't a concern anymore, providing full paths as suggested is the best solution.

Note: This will probably be a XL PR since it impacts application_types, zz_* and a number of methods 
in kustomize.

**Comment by yanniszark** [comment]
@jlewi is this really fixed in https://github.com/kubeflow/kubeflow/pull/3548 ?
It seems to me that you still can't specify full path to components:
eg. `common/centraldashboard` vs `centraldashboard`

**Comment by jlewi** [comment]
@yanniszark not fixed; I meant to merely say it was related.

**Comment by jlewi** [comment]
I considered a variety of temporary fixes. For example, I considered prepending a "/" to the componentName to indicate it was an actual path and discovery logic shouldn't apply.

My conclusion though is that this would be roughly as much work as updating the code to use KfDef.Spec.Applications and be more brittle.

So I will take a stab at 

1. defining KfDef.Spec.Applications
1. Updating the code to use it
1. Providing a backFill method to compute .Applications from components, packages, and componentParams so that existing manifests continue to work.

**Comment by yanniszark** [comment]
@jlewi sounds good, I just got confused at this part:

> Providing a backFill method to compute .Applications from components, packages, and componentParams so that existing manifests continue to work.

Do you mean that you will backfill using the existing broken discovery logic?

**Comment by jlewi** [comment]
Yeas. If KfDef doesn't have Applications defined; but still uses components, packages, componentParams then we will use the existing logic to locate the packages. We will then convert it to Applications and update KfDef.

**Comment by jlewi** [comment]
I have fixes pending
First PR is #3594 but more PRs will be coming based on #3534.

I don't think we will block 0.6.0 on this fix. So I'm going to downgrade this to P1 since its not-blocking.

The fix will be based on using the new KfDef.Applications which allows each Application to specify the path of the application.

So to get the fix people will need to migrate to the new spec.

**Comment by jlewi** [comment]
All of the changes to support KfDef.Applications are merged.  They should also be included in the v0.6.0 release.

Remaining work would be to upgrade the specs and tests to use it. We should file additional bugs for that.
